### PR TITLE
Decorations: respect earlier decoration negotiation results

### LINKED ISSUE #297: Inconsistency between CSD/SSD reported to client, and actually rendered
As mentioned in https://github.com/labwc/labwc/issues/296#issuecomment-1100699231, setting `<decoration>client</decoration>` doesn't actually enable CSDs in foot, but does in Alacritty. What gives?

The difference is how these two applications use the xdg-decoraction protocol. This protocol has two methods: `set_mode()` and `unset_mode()`.

Alacritty uses `unset_mode()`, while foot uses `set_mode()`.

`unset_mode()`:

> Unset the toplevel surface decoration mode. This informs the compositor that the client doesn't prefer a particular decoration mode.

`set_mode()`:

> Set the toplevel surface decoration mode. This informs the compositor that the client prefers the provided decoration mode 
After requesting a decoration mode, the compositor will respond by emitting an xdg_surface.configure event. The client should then update its content, drawing it without decorations if the received mode is server-side decorations. The client must also acknowledge the configure when committing the new content (see xdg_surface.ack_configure).
> 
> **The compositor can decide not to use the client's mode and enforce a different mode instead.**

Key words here are **prefers** and **can enforce a different mode instead**.

What I suspect is happening is that labwc receives foot's `set_mode()` request, where foot says it **prefers** SSDs, and labwc replies with "sure ok, no problem" (i.e. foot receives `ZXDG_TOPLEVEL_DECORATION_V1_MODE_SERVER_SIDE` in the xdg toplevel decoration `configure` event - this has been verified, see attached ` WAYLAND_DEBUG` logs).

But when rendering the foot window, labwc just looks at its `<decoration>` setting and decides to not draw any SSDs.

`WAYLAND_DEBUG` logs (`grep decoration`  - full logs are attached):

Alacritty
```
[1604551.033] wl_registry@2.global(15, "zxdg_decoration_manager_v1", 1)
[1604551.049]  -> wl_registry@2.bind(15, "zxdg_decoration_manager_v1", 1, new id [unknown]@8)
[1604551.068] wl_registry@2.global(16, "org_kde_kwin_server_decoration_manager", 1)
[1604607.326]  -> zxdg_decoration_manager_v1@8.get_toplevel_decoration(new id zxdg_toplevel_decoration_v1@22, xdg_toplevel@19)
[1604607.352]  -> zxdg_toplevel_decoration_v1@22.unset_mode()
[1604615.769] zxdg_toplevel_decoration_v1@22.configure(1)
[1604621.488] wl_registry@38.global(15, "zxdg_decoration_manager_v1", 1)
[1604621.501] wl_registry@38.global(16, "org_kde_kwin_server_decoration_manager", 1)
[1604728.868] wl_registry@39.global(15, "zxdg_decoration_manager_v1", 1)
[1604728.875] wl_registry@39.global(16, "org_kde_kwin_server_decoration_manager", 1)
```

Foot
```
[1614107.530] wl_registry@2.global(15, "zxdg_decoration_manager_v1", 1)
[1614107.542]  -> wl_registry@2.bind(15, "zxdg_decoration_manager_v1", 1, new id [unknown]@14)
[1614107.558] wl_registry@2.global(16, "org_kde_kwin_server_decoration_manager", 1)
[1614110.118]  -> zxdg_decoration_manager_v1@14.get_toplevel_decoration(new id zxdg_toplevel_decoration_v1@23, xdg_toplevel@22)
info: wayland.c:1453: requesting SSD decorations
[1614110.139]  -> zxdg_toplevel_decoration_v1@23.set_mode(2)
[1614181.568] zxdg_toplevel_decoration_v1@23.configure(2)
info: wayland.c:818: using SSD decorations
[1614752.888]  -> zxdg_toplevel_decoration_v1@23.destroy()
[1614754.351]  -> zxdg_decoration_manager_v1@14.destroy()
```

[foot.log](https://github.com/labwc/labwc/files/8500334/foot.log)
[alacritty.log](https://github.com/labwc/labwc/files/8500335/alacritty.log)

labwc: 0.5.0-100-g6b4d9b9
Alacritty: 0.10.1
Foot: latest master

### LINKED ISSUE #831: Element (Electron based) not showing decorations when CSD is enabled in rc.xml
Certain programs like Element, MPV and PPSSPP (all of them wayland native) don't show any decorations when CSD is enabled. (`<core><decoration>client</decoration></core>`)

I managed to "fix" it with a -really- nasty hack in `xdg-deco.c` ; it works but could bring other problems I believe. Here's the patch: https://github.com/umksan/labwc/commit/47f4b67f2c0b01d921e39e6ddd07d5b10b128fd9

Since Element uses Electron I have tried using `--enable-features=WaylandWindowDecorations` but decorations didn't appear either.

On the other hand, a retro gaming client Fightcade (which uses Electron but runs on Xwayland) is able to show SSD while CSD is enabled in rc.xml.

----

Thanks to @Consolatis for helping me to debug behaviour correctly. I tested multiple programs. All of them wayland native.

MPV and Element return this:

```
00:01:56.256 [../src/xdg-deco.c:31] Client requested server decorations
00:01:56.256 [../src/xdg-deco.c:39] We set decorations to server
```

PPSSPP:

```
00:00:01.932 [../src/xdg-deco.c:31] Client requested None decorations
00:00:01.932 [../src/xdg-deco.c:39] We set decorations to client
00:00:01.932 [../src/xdg-deco.c:31] Client requested server decorations
00:00:01.932 [../src/xdg-deco.c:39] We set decorations to server
```

### PR DESCRIPTION
**Pull request by Consolatis**
Before this patch, setting `<decoration>` to `client` would cause applications
which prefer server side decorations to not have any decorations at all. This
patch fixes it by respecting the result of earlier negotiations via the
xdg-decoration-unstable-v1 protocol.

This patch also slightly changes the behavior of the decoration negotiation
depending on the user configured decoration settings in rc.xml:
- If set to `server`: application preferences are ignored
- If set to `client`: application settings are honored. This makes sure clients
  that don't actually implement any kind of client side decorations (and thus
  request server side decorations) will not end up without any decorations.

Fixes #297
Fixes #831

---
This is a draft for now until it received some testing to prevent regressions.

### CODE REVIEW

**Comment by Consolatis** [comment]
```
<Consolatis> hm... chromium plainly seems to ignore when we tell it to use server side decorations
<Consolatis> it asks for client side, gets told to use server side and yet still uses client side ones
<Consolatis> it can be switched via config to use "system titlebar and borders" though
<Consolatis> i guess we are back to the beginning of the question what should happen when a users sets rc.xml to "server": overwrite the preferences of clients or accept them
```

**Comment by Consolatis** [comment]
Squashed the commits, this PR now only fixes the "no decorations at all when configured `client` decorations" case and moves `xdg-deco.c` to `src/decorations/`. This seems ready to go.

File: src/server.c [hunk 1]
```diff
@@ -310,26 +312,8 @@ server_init(struct server *server)
 		&server->new_xdg_surface);
 
 	/* Disable CSD */
```
Comment (johanmalm) [inline src/server.c:314]: Remove comment 'Disable CSD'??

File: src/xdg.c [hunk 2]
```diff
@@ -48,21 +48,19 @@ handle_new_xdg_popup(struct wl_listener *listener, void *data)
...
-
-	/*
-	 * Some XDG shells refuse to disable CSD in which case their
-	 * geometry.{x,y} seems to be greater than zero. We filter on that
-	 * on the assumption that this will remain true.
-	 */
-	struct wlr_xdg_surface_state *current =
-		&xdg_surface_from_view(view)->current;
-	if (current->geometry.x || current->geometry.y) {
+	/* view->ssd_preference is set by the kde or xdg decoration variant */
```
Comment (johanmalm) [inline src/xdg.c:51]: Spell out `xdg-decoration` or `kde-server-decoration`

File: src/xdg.c [hunk 3]
```diff
@@ -526,6 +524,7 @@ xdg_surface_new(struct wl_listener *listener, void *data)
 	view->type = LAB_XDG_SHELL_VIEW;
 	view->impl = &xdg_toplevel_view_impl;
 	xdg_toplevel_view->xdg_surface = xdg_surface;
+	view->ssd_preference = LAB_SSD_PREF_CLIENT;
```
Comment (johanmalm) [inline src/xdg.c:527]: Should we explain why we set `ssd_preference` to client here? I get it following IRC discussion, but it's pretty non-intuitive.

File: src/server.c (hunk 1 above)
Comment (Consolatis) [reply src/server.c]: removed

File: src/xdg.c (hunk 2 above)
Comment (Consolatis) [reply src/xdg.c]: I've used the `.c` files instead but happy to change to the protocol names if desired.

File: src/xdg.c (hunk 3 above)
Comment (Consolatis) [reply src/xdg.c]: Added some more comments

**Comment by johanmalm** [comment]
LGTM.
…and looks a well crafted/encapsulated solution. Really nice.
Happy to merge now.
Or I can do some testing in the morning and merge then.

**Comment by Consolatis** [comment]
> Or I can do some testing in the morning and merge then.

That would be great. It also provides for some more time to detect regressions.

**Comment by johanmalm** [comment]
Have tested with `<decoration>` set to both client|server with a range of applications. Seems to work well. Will go for merge.

**Pull request merged**